ls1intum/Artemis
# `Development`: Enable static code analysis in local continuous integration for Java

### Checklist
#### General


#### Server

### Motivation and Context


At the moment, the LocalCI + LocalVC setup doesn't support Static Code Analysis, which is an important part of the automatic assessment of programming exercises in Artemis.

### Description

Static Code Analysis was enabled. The SCA configuration files for Java were already given in the `resources` directory. I had to add the path where to find the result files in `LocalCIBuildJobExecutionService` and also add the commands to the script for Gradle and Maven. Lastly, by adapting and refactoring the code from https://github.com/ls1intum/static-code-analysis-parser, we can now parse the result files for various languages and give the user feedback for their submissions.
If we enable SCA for more programming languages in the future, we will already have the corresponding parsers, too.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Programming Exercise with Static Code Analysis enabled

Create a Java Exercise with static code analysis enabled. Add some code that should trigger some warnings, for example, an uninitialised variable. Submit the code and check if you get a SCA warning.

### Testserver States
> [!NOTE]
> These badges show the state of the test servers.
> Green = Currently available, Red = Currently locked

![](https://byob.yarr.is/ls1intum/Artemis/artemis-test1)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test2)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test3)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test4)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test5)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test6)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test9)

### Review Progress






#### Performance Review
#### Code Review
#### Manual Tests


![image](https://github.com/ls1intum/Artemis/assets/120647937/c1beb6b1-59c4-42db-ae2f-c1e78ce35cb0)
